File: matlab/benchmark/check_publication_perf_gate.m
function summary = check_publication_perf_gate(varargin)
%CHECK_PUBLICATION_PERF_GATE Compare two publication CSVs with slowdown gate.
%
%   summary = check_publication_perf_gate()
%   summary = check_publication_perf_gate(baseline_csv, candidate_csv, max_slowdown)

repo_root = fileparts(fileparts(fileparts(mfilename('fullpath'))));
def_csv = fullfile(repo_root, 'matlab', 'benchmark', 'results', 'publication', 'publication_timings.csv');

if nargin >= 1
    baseline_csv = varargin{1};
else
    baseline_csv = def_csv;
end
if nargin >= 2
    candidate_csv = varargin{2};
else
    candidate_csv = def_csv;
end
if nargin >= 3
    max_slowdown = varargin{3};
else
    max_slowdown = 0.05;
end

if ~(isnumeric(max_slowdown) && isscalar(max_slowdown) && isfinite(max_slowdown) && max_slowdown >= 0)
    error('check_publication_perf_gate:InvalidThreshold', ...
        'max_slowdown must be a nonnegative finite scalar.');
end

base = normalize_text_columns(readtable(baseline_csv));
cand = normalize_text_columns(readtable(candidate_csv));

methods = ["bsqr_full", "bsqr_rinv"];
base = base(ismember(base.method, methods), :);
cand = cand(ismember(cand.method, methods), :);

key_cols = {'family', 'regime', 'm', 'n', 'aspect', 'seed', 'method'};
if ismember('bench_surface', base.Properties.VariableNames) || ismember('bench_surface', cand.Properties.VariableNames)
    if ~ismember('bench_surface', base.Properties.VariableNames)
        base.bench_surface = repmat("factor_only", height(base), 1);
    end
    if ~ismember('bench_surface', cand.Properties.VariableNames)
        cand.bench_surface = repmat("factor_only", height(cand), 1);
    end
    key_cols = [key_cols, {'bench_surface'}];
end
base_t = base(:, [key_cols, {'tmed_s'}]);
base_t.Properties.VariableNames{'tmed_s'} = 'tmed_base';
cand_t = cand(:, [key_cols, {'tmed_s'}]);
cand_t.Properties.VariableNames{'tmed_s'} = 'tmed_candidate';

joined = innerjoin(base_t, cand_t, 'Keys', key_cols);
if isempty(joined)
    error('check_publication_perf_gate:NoComparableRows', ...
        'No overlapping method/case rows were found between baseline and candidate CSVs.');
end

joined.slowdown = joined.tmed_candidate ./ joined.tmed_base - 1.0;
joined = sortrows(joined, {'method', 'family', 'regime', 'm', 'n', 'aspect', 'seed'});

viol = joined(isfinite(joined.slowdown) & joined.slowdown > max_slowdown, :);
summary = summarize_slowdown(joined);
summary.max_slowdown = max_slowdown;
summary.num_compared = height(joined);
summary.num_violations = height(viol);
summary.violations = viol;

fprintf('Performance gate compared %d rows (threshold %.2f%% slowdown).\n', ...
    summary.num_compared, 100 * max_slowdown);
disp(summary.by_method);
if ~isempty(viol)
    fprintf('Found %d violating rows.\n', height(viol));
    disp(viol(:, {'method', 'family', 'regime', 'm', 'n', 'aspect', 'seed', 'slowdown'}));
    error('check_publication_perf_gate:Regression', ...
        'Performance gate failed: slowdown exceeded threshold.');
end
fprintf('Performance gate passed.\n');
end

function out = summarize_slowdown(joined)
methods = unique(joined.method);
med = zeros(numel(methods), 1);
mx = zeros(numel(methods), 1);
for i = 1:numel(methods)
    mask = joined.method == methods(i);
    vals = joined.slowdown(mask);
    med(i) = median(vals);
    mx(i) = max(vals);
end
out = struct();
out.by_method = table(methods, med, mx, ...
    'VariableNames', {'method', 'median_slowdown', 'max_slowdown'});
end

function rows = normalize_text_columns(rows)
text_cols = {'family', 'regime', 'method', 'run_id', 'timestamp', 'bench_surface'};
for i = 1:numel(text_cols)
    c = text_cols{i};
    if ~ismember(c, rows.Properties.VariableNames)
        continue;
    end
    v = rows.(c);
    if iscellstr(v)
        rows.(c) = string(v);
    elseif ischar(v)
        rows.(c) = string(cellstr(v));
    elseif iscategorical(v)
        rows.(c) = string(v);
    end
end
end

File: julia/benchmark/check_publication_perf_gate.jl
#!/usr/bin/env julia

using DelimitedFiles
using Printf
using Statistics

const DEFAULT_INPUT = joinpath(@__DIR__, "results", "publication", "publication_timings.csv")
const METHODS = Set(["bsqr_full", "bsqr_rinv"])

_as_string(x) = String(x)
_as_int(x) = x isa Int ? x : parse(Int, _as_string(x))
_as_float(x) = x isa AbstractFloat ? Float64(x) : parse(Float64, _as_string(x))

function _load(csv_path::String)
    isfile(csv_path) || error("CSV not found: $csv_path")
    data, raw_header = readdlm(csv_path, ',', header = true)
    header = _as_string.(vec(raw_header))
    idx = Dict{String,Int}(h => i for (i, h) in enumerate(header))
    required = ["family", "regime", "m", "n", "aspect", "seed", "method", "tmed_s"]
    for c in required
        haskey(idx, c) || error("Missing required column '$c' in $csv_path")
    end

    out = Dict{Tuple{String,String,Int,Int,Float64,Int,String},Float64}()
    for i in 1:size(data, 1)
        row = view(data, i, :)
        method = _as_string(row[idx["method"]])
        method in METHODS || continue
        key = (
            _as_string(row[idx["family"]]),
            _as_string(row[idx["regime"]]),
            _as_int(row[idx["m"]]),
            _as_int(row[idx["n"]]),
            _as_float(row[idx["aspect"]]),
            _as_int(row[idx["seed"]]),
            method,
        )
        out[key] = _as_float(row[idx["tmed_s"]])
    end
    return out
end

function main()
    baseline_csv = length(ARGS) >= 1 ? ARGS[1] : DEFAULT_INPUT
    candidate_csv = length(ARGS) >= 2 ? ARGS[2] : DEFAULT_INPUT
    max_slowdown = length(ARGS) >= 3 ? parse(Float64, ARGS[3]) : 0.05
    # Cells whose baseline median is below this floor cannot carry a slowdown
    # signal across runs (timer/scheduler noise dominates sub-ms cases); they
    # are reported in the aggregates but exempt from the gate.
    min_tmed = length(ARGS) >= 4 ? parse(Float64, ARGS[4]) : 0.0
    max_slowdown >= 0 || error("max_slowdown must be >= 0")
    min_tmed >= 0 || error("min_tmed must be >= 0")

    base = _load(baseline_csv)
    cand = _load(candidate_csv)

    common_keys = sort!(collect(intersect(Set(Base.keys(base)), Set(Base.keys(cand)))))
    isempty(common_keys) && error("No overlapping rows between baseline and candidate.")

    by_method = Dict("bsqr_full" => Float64[], "bsqr_rinv" => Float64[])
    violations = Tuple{Tuple{String,String,Int,Int,Float64,Int,String},Float64}[]
    ngated = 0
    for k in common_keys
        tb = base[k]
        tc = cand[k]
        tb > 0 || error("Non-positive baseline tmed for key=$k")
        slowdown = tc / tb - 1.0
        push!(by_method[k[7]], slowdown)
        tb >= min_tmed || continue
        ngated += 1
        if isfinite(slowdown) && slowdown > max_slowdown
            push!(violations, (k, slowdown))
        end
    end

    println("Performance gate compared $(length(common_keys)) rows, $ngated gated (threshold $(round(100 * max_slowdown, digits=2))% slowdown, min_tmed $(min_tmed)s).")
    println("| method | median_slowdown | max_slowdown |")
    println("|---|---:|---:|")
    for m in ["bsqr_full", "bsqr_rinv"]
        vals = by_method[m]
        isempty(vals) && continue
        @printf("| %s | %.6f | %.6f |\n", m, median(vals), maximum(vals))
    end

    if !isempty(violations)
        println("\nGate failed with $(length(violations)) violating rows.")
        for (k, s) in violations
            @printf("%s slowdown=%.6f\n", repr(k), s)
        end
        error("Performance gate failed: slowdown exceeded threshold.")
    end

    println("Performance gate passed.")
end

main()
